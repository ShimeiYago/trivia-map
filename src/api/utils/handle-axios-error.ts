import { AxiosError } from 'axios';

export function handleAxiosError(axiosError: AxiosError): ApiError {
  if (axiosError.response) {
    // when API is active but response includes error
    const error: ApiError = {
      status: axiosError.response.status,
      data: axiosError.response.data,
      errorMsg: axiosError.message,
    };
    return error;
  } else {
    // when API is not active
    const error: ApiError = {
      status: 500,
      data: {},
      errorMsg: axiosError.message,
    };
    return error;
  }
}

export type ApiError = {
  status: number;
  data: unknown;
  errorMsg: string;
};
