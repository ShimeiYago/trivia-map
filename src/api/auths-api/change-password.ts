import { BASE_URL } from 'constant';
import { AxiosError } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function changePassword(
  password1: string,
  password2: string,
): Promise<void> {
  const axiosInstance = getAxiosInstance({}, {});

  const requestData: ChangePasswordRequest = {
    password1: password1,
    password2: password2,
  };

  try {
    // TODO: Set reasonable timeout
    await axiosInstance.post(`${BASE_URL}/auths/password/change`, requestData);
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError<ValidationError>(axiosError);
  }
}

export type ChangePasswordRequest = {
  password1: string;
  password2: string;
};

export type ValidationError = {
  password1?: string[];
  password2?: string[];
};
