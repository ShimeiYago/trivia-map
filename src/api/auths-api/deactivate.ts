import { API_TIMEOUT, BASE_URL } from 'constant';
import { AxiosError } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';

export async function deactivate(): Promise<void> {
  const axiosInstance = getAxiosInstance({ timeout: API_TIMEOUT.long }, {});

  try {
    await axiosInstance.put(`${BASE_URL}/auths/deactivate/`);
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}
