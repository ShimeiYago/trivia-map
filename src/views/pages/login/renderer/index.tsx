import { ADMIN_LINK } from 'constant/links';
import React from 'react';
import { Navigate } from 'react-router-dom';
import { sleep } from 'utils/sleep';
import { AuthForms } from 'views/components/organisms/auth-forms';

export class Renderer extends React.Component<unknown, State> {
  state: State = {
    redirect: false,
  };

  render() {
    if (this.state.redirect) {
      return <Navigate to={ADMIN_LINK} />;
    }

    return (
      <AuthForms
        initialMode="login"
        onLoginSucceed={this.redirectAfterSeveralSeconds}
      />
    );
  }

  protected redirectAfterSeveralSeconds = async () => {
    await sleep(1000);
    this.setState({
      redirect: true,
    });
  };
}

export type State = {
  redirect: boolean;
};
