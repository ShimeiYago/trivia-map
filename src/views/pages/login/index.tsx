import { Renderer } from './renderer';

export const Login = Renderer;
