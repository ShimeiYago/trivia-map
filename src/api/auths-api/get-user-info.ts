import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { User } from 'types/user';
import { mockGetUserInfoResponse } from 'api/mock/auths-response/get-user-info';

export async function getUserInfo(): Promise<User> {
  const axiosInstance = getAxiosInstance({}, mockGetUserInfoResponse);

  try {
    // TODO: Set reasonable timeout
    const res: AxiosResponse<User> = await axiosInstance.get(`${BASE_URL}/auths/user`);
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError(axiosError);
  }
}
