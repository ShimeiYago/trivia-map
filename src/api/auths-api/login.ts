import { BASE_URL } from 'constant';
import { AxiosError, AxiosResponse } from 'axios';
import { handleAxiosError } from '../utils/handle-axios-error';
import { getAxiosInstance } from 'api/utils/get-axios-instance';
import { mockLoginResponse } from 'api/mock/auths-response/login';
import { User } from 'types/user';

export async function login(
  email: string,
  password: string,
): Promise<LoginResponse> {
  const axiosInstance = getAxiosInstance({}, mockLoginResponse);

  const requestData: LoginRequest = {
    email: email,
    password: password,
  };

  try {
    // TODO: Set reasonable timeout
    const res: AxiosResponse<LoginResponse> = await axiosInstance.post(
      `${BASE_URL}/auths/login`,
      requestData,
    );
    return res.data;
  } catch (error) {
    const axiosError = error as AxiosError;
    throw handleAxiosError<ValidationError>(axiosError);
  }
}

export type LoginRequest = {
  email: string;
  password: string;
};

export type LoginResponse = {
  access_token: string;
  refresh_token: string;
  user: User;
};

export type ValidationError = {
  email?: string[];
  password?: string[];
  non_field_errors?: string[];
};
