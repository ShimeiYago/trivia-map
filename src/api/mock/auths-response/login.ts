import { LoginResponse } from 'api/auths-api/login';

export const mockLoginResponse: LoginResponse = {
  access_token: 'xxx',
  refresh_token: 'xxx',
  user: {
    userId: 1,
    email: 'user@example.com',
    nickname: 'Axel',
    icon: 'https://disneyparkstory.com/wp-content/uploads/2017/09/apple-touch-icon.png',
  },
  access_token_expiration: '2023-10-05T01:31:10.861050+09:00',
  refresh_token_expiration: '2023-10-05T01:31:10.861050+09:00',
};
