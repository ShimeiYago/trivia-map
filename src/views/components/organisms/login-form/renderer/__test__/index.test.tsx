import { shallow, ShallowWrapper } from 'enzyme';
import { Renderer, Props, State } from '..';
import * as LoginModule from 'api/auths-api/login';
import { mockLoginResponse } from 'api/mock/auths-response/login';
import { ApiError } from 'api/utils/handle-axios-error';

let wrapper: ShallowWrapper<Props, State, Renderer>;

let loginSpy: jest.SpyInstance;

const basicProps: Props = {
  logginingInState: 'waiting',
  loginSuccess: jest.fn(),
};

describe('Shallow Snapshot Tests', () => {
  beforeEach(() => {
    wrapper = shallow(<Renderer {...basicProps} />);
  });

  it('basic', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('error case', () => {
    wrapper.setState({
      errorTitle: 'error title',
      errorMessages: ['error1', 'error2'],
      formError: {
        email: ['email is invalid'],
        password: ['password is invalid'],
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('already logged in', () => {
    wrapper.setProps({
      logginingInState: 'success',
    });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('handleChangeTextField', () => {
  it('should set email', () => {
    wrapper = shallow(<Renderer {...basicProps} />);
    const instance = wrapper.instance();

    const event = {
      target: {
        value: 'xxx@xxx.com',
      },
    } as React.ChangeEvent<HTMLInputElement>;

    instance['handleChangeTextField']('email')(event);

    expect(instance.state.email).toBe('xxx@xxx.com');
  });

  it('should set password', async () => {
    wrapper = shallow(<Renderer {...basicProps} />);
    const instance = wrapper.instance();

    const event = {
      target: {
        value: 'xxxxx',
      },
    } as React.ChangeEvent<HTMLInputElement>;

    instance['handleChangeTextField']('password')(event);

    expect(instance.state.password).toBe('xxxxx');
  });
});

describe('handleClickLogin', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    loginSpy = jest.spyOn(LoginModule, 'login');
  });

  it('should set localLoadingState success when api succeed', async () => {
    loginSpy.mockResolvedValue(mockLoginResponse);

    wrapper = shallow(<Renderer {...basicProps} />);
    const instance = wrapper.instance();

    await instance['handleClickLogin']();

    expect(instance.state.localLoadingState).toBe('success');
  });

  it('should set form error when api have validation error', async () => {
    const apiError: ApiError<LoginModule.ValidationError> = {
      status: 400,
      data: {
        email: ['email is invalid'],
        password: ['password is invalid'],
        non_field_errors: [
          'another error',
          'Must include "email" and "password".',
          'E-mail is not verified.',
        ],
      },
      errorMsg: '400 request is invalid',
    };
    loginSpy.mockRejectedValue(apiError);

    wrapper = shallow(<Renderer {...basicProps} />);
    const instance = wrapper.instance();

    await instance['handleClickLogin']();

    expect(instance.state.formError).toEqual({
      email: ['email is invalid'],
      password: ['password is invalid'],
    });
  });

  it('should set localLoadingState error when api fail', async () => {
    loginSpy.mockRejectedValue(new Error());

    wrapper = shallow(<Renderer {...basicProps} />);
    const instance = wrapper.instance();

    await instance['handleClickLogin']();

    expect(instance.state.localLoadingState).toBe('error');
  });
});
