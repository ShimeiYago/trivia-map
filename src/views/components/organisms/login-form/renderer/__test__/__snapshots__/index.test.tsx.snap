// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shallow Snapshot Tests already logged in 1`] = `
<ForwardRef(Container)
  component="main"
  maxWidth="xs"
>
  <ForwardRef(Box)>
    <ForwardRef(Typography)
      align="center"
      component="h1"
      variant="h5"
    >
      ログイン
    </ForwardRef(Typography)>
    <ForwardRef(Box)
      component="form"
      noValidate={true}
      sx={
        Object {
          "mt": 1,
        }
      }
    >
      <ForwardRef(TextField)
        autoComplete="email"
        autoFocus={true}
        disabled={true}
        error={false}
        fullWidth={true}
        id="email"
        label="メールアドレス"
        margin="normal"
        name="email"
      />
      <ForwardRef(TextField)
        autoComplete="current-password"
        disabled={true}
        error={false}
        fullWidth={true}
        id="password"
        label="パスワード"
        margin="normal"
        name="password"
        type="password"
      />
      <ForwardRef(LoadingButton)
        disabled={true}
        fullWidth={true}
        loading={false}
        onClick={[Function]}
        sx={
          Object {
            "mb": 2,
            "mt": 3,
          }
        }
        variant="contained"
      >
        ログイン
      </ForwardRef(LoadingButton)>
      <ForwardRef(Grid)
        container={true}
      >
        <ForwardRef(Grid)
          item={true}
          xs={true}
        >
          <ForwardRef(Button)
            variant="text"
          >
            パスワードを忘れた場合
          </ForwardRef(Button)>
        </ForwardRef(Grid)>
        <ForwardRef(Grid)
          item={true}
        >
          <ForwardRef(Button)
            variant="text"
          >
            アカウント作成
          </ForwardRef(Button)>
        </ForwardRef(Grid)>
      </ForwardRef(Grid)>
    </ForwardRef(Box)>
  </ForwardRef(Box)>
</ForwardRef(Container)>
`;

exports[`Shallow Snapshot Tests basic 1`] = `
<ForwardRef(Container)
  component="main"
  maxWidth="xs"
>
  <ForwardRef(Box)>
    <ForwardRef(Typography)
      align="center"
      component="h1"
      variant="h5"
    >
      ログイン
    </ForwardRef(Typography)>
    <ForwardRef(Box)
      component="form"
      noValidate={true}
      sx={
        Object {
          "mt": 1,
        }
      }
    >
      <ForwardRef(TextField)
        autoComplete="email"
        autoFocus={true}
        disabled={false}
        error={false}
        fullWidth={true}
        id="email"
        label="メールアドレス"
        margin="normal"
        name="email"
      />
      <ForwardRef(TextField)
        autoComplete="current-password"
        disabled={false}
        error={false}
        fullWidth={true}
        id="password"
        label="パスワード"
        margin="normal"
        name="password"
        type="password"
      />
      <ForwardRef(LoadingButton)
        disabled={false}
        fullWidth={true}
        loading={false}
        onClick={[Function]}
        sx={
          Object {
            "mb": 2,
            "mt": 3,
          }
        }
        variant="contained"
      >
        ログイン
      </ForwardRef(LoadingButton)>
      <ForwardRef(Grid)
        container={true}
      >
        <ForwardRef(Grid)
          item={true}
          xs={true}
        >
          <ForwardRef(Button)
            variant="text"
          >
            パスワードを忘れた場合
          </ForwardRef(Button)>
        </ForwardRef(Grid)>
        <ForwardRef(Grid)
          item={true}
        >
          <ForwardRef(Button)
            variant="text"
          >
            アカウント作成
          </ForwardRef(Button)>
        </ForwardRef(Grid)>
      </ForwardRef(Grid)>
    </ForwardRef(Box)>
  </ForwardRef(Box)>
</ForwardRef(Container)>
`;

exports[`Shallow Snapshot Tests error case 1`] = `
<ForwardRef(Container)
  component="main"
  maxWidth="xs"
>
  <ForwardRef(Box)>
    <ForwardRef(Typography)
      align="center"
      component="h1"
      variant="h5"
    >
      ログイン
    </ForwardRef(Typography)>
    <ForwardRef(Alert)
      severity="error"
    >
      <ForwardRef(AlertTitle)>
        error title
      </ForwardRef(AlertTitle)>
      <ul>
        <li
          key="error-0"
        >
          error1
        </li>
        <li
          key="error-1"
        >
          error2
        </li>
      </ul>
    </ForwardRef(Alert)>
    <ForwardRef(Box)
      component="form"
      noValidate={true}
      sx={
        Object {
          "mt": 1,
        }
      }
    >
      <ForwardRef(TextField)
        autoComplete="email"
        autoFocus={true}
        disabled={false}
        error={true}
        fullWidth={true}
        helperText={
          Array [
            "email is invalid",
          ]
        }
        id="email"
        label="メールアドレス"
        margin="normal"
        name="email"
      />
      <ForwardRef(TextField)
        autoComplete="current-password"
        disabled={false}
        error={true}
        fullWidth={true}
        helperText={
          Array [
            "password is invalid",
          ]
        }
        id="password"
        label="パスワード"
        margin="normal"
        name="password"
        type="password"
      />
      <ForwardRef(LoadingButton)
        disabled={false}
        fullWidth={true}
        loading={false}
        onClick={[Function]}
        sx={
          Object {
            "mb": 2,
            "mt": 3,
          }
        }
        variant="contained"
      >
        ログイン
      </ForwardRef(LoadingButton)>
      <ForwardRef(Grid)
        container={true}
      >
        <ForwardRef(Grid)
          item={true}
          xs={true}
        >
          <ForwardRef(Button)
            variant="text"
          >
            パスワードを忘れた場合
          </ForwardRef(Button)>
        </ForwardRef(Grid)>
        <ForwardRef(Grid)
          item={true}
        >
          <ForwardRef(Button)
            variant="text"
          >
            アカウント作成
          </ForwardRef(Button)>
        </ForwardRef(Grid)>
      </ForwardRef(Grid)>
    </ForwardRef(Box)>
  </ForwardRef(Box)>
</ForwardRef(Container)>
`;
